pytential/symbolic/stokes.py

```diff
@@ -55,28 +56,79 @@ class StokesletWrapper:
+            raise ValueError("unsupported dimension given to StokesletWrapper")
 
-        elif dim == 3:
-            self.kernel_dict = {
-                        (2, 0, 0): StokesletKernel(dim=3, icomp=0, jcomp=0),
-                        (1, 1, 0): StokesletKernel(dim=3, icomp=0, jcomp=1),
-                        (1, 0, 1): StokesletKernel(dim=3, icomp=0, jcomp=2),
-                        (0, 2, 0): StokesletKernel(dim=3, icomp=1, jcomp=1),
-                        (0, 1, 1): StokesletKernel(dim=3, icomp=1, jcomp=2),
-                        (0, 0, 2): StokesletKernel(dim=3, icomp=2, jcomp=2)
-                               }
+        self.kernel_dict = {}
 
-        else:
-            raise ValueError("unsupported dimension given to StokesletWrapper")
+        self.base_kernel = BiharmonicKernel(dim=dim)
+
+        if not use_biharmonic:
+            for i in range(dim):
+                for j in range(i, dim):
+                    self.kernel_dict[(i, j)] = StokesletKernel(dim=dim, icomp=i,
+                                                               jcomp=j)
+
+            for i in range(dim):
+                for j in range(i):
+                    self.kernel_dict[(i, j)] = self.kernel_dict[(j, i)]
+
+    def _get_stokeslet_int_g(self, icomp, jcomp, density, mu_sym, qbx_forced_limit,
+            deriv_dirs):
+
+        """
+        Returns the Integral of the Stokeslet kernel given by `icomp` and `jcomp`
+        and its derivatives.
+        Note that,
+
+            Stokeslet(icomp, jcomp)
+                = -1/mu_sym d/(d x_icomp) d/(d x_jcomp) BiharmonicKernel
+
+        for icomp != jcomp. Also note that for 3D,
+
+            Stokeslet(i, i)
+                = -1/mu_sym sum(-d^2/(dx_j^2 BiharmonicKernel for j != i).
+
+        For 2D,
+
+            Stokeslet(i, i)
+                = -1/mu_sym (sum(d^2/dx_j^2 BiharmonicKernel for j != i)
+                             -3/(8*pi)).
+        """
+
+        if not self.use_biharmonic:
+            knl = self.kernel_dict[(icomp, jcomp)]
+            for deriv_dir in deriv_dirs:
+                knl = AxisTargetDerivative(deriv_dir, knl)
+            return sym.IntG(knl, density,
+                    qbx_forced_limit=qbx_forced_limit, mu=mu_sym)
+
+
+        def func(knl, deriv_dirs):
+            for deriv_dir in deriv_dirs:
+                knl = AxisTargetDerivative(deriv_dir, knl)
+
+            res = sym.IntG(knl, density,
+                    qbx_forced_limit=qbx_forced_limit)
+            return res
+
+        mult = -1 / mu_sym
+
+        if icomp != jcomp:
+            return mult * func(self.base_kernel,
+                deriv_dirs=(deriv_dirs + [icomp, jcomp]))
+
+        const = 0
+        if self.dim == 2 and not deriv_dirs:
+            from math import pi
+            const = -3/(8*pi)
+
+        return -mult * (sum(func(self.base_kernel,
+                deriv_dirs=(deriv_dirs + [i, i])) for i in range(self.dim)
+                if i != icomp) + const * density)
```

Comment:
@inducer, because of this `const*density` I get the following error,

```
test/test_stokes.py:229: in run_exterior_stokes_2d
    omega=omega)
pytential/symbolic/execution.py:1053: in __call__
    return self.eval(kwargs, array_context=array_context)
pytential/symbolic/execution.py:1028: in eval
    return self.code.execute(exec_mapper)
pytential/symbolic/compiler.py:387: in execute
    insn, exec_mapper.bound_expr, exec_mapper))
pytential/symbolic/execution.py:274: in exec_assign
    for name, expr in zip(insn.names, insn.exprs)]
pytential/symbolic/execution.py:274: in <listcomp>
    for name, expr in zip(insn.names, insn.exprs)]
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
../pymbolic/pymbolic/mapper/evaluator.py:92: in map_sum
    return sum(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/evaluator.py:92: in <genexpr>
    return sum(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
../pymbolic/pymbolic/mapper/evaluator.py:96: in map_product
    return product(self.rec(child) for child in expr.children)
../../miniconda3/lib/python3.6/site-packages/pytools/__init__.py:978: in product
    return reduce(mul, iterable, 1)
../pymbolic/pymbolic/mapper/evaluator.py:96: in <genexpr>
    return product(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
../pymbolic/pymbolic/mapper/evaluator.py:99: in map_quotient
    return self.rec(expr.numerator) / self.rec(expr.denominator)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
pytential/symbolic/execution.py:118: in map_node_sum
    for grp_ary in self.rec(expr.operand))
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
../pymbolic/pymbolic/mapper/evaluator.py:96: in map_product
    return product(self.rec(child) for child in expr.children)
../../miniconda3/lib/python3.6/site-packages/pytools/__init__.py:978: in product
    return reduce(mul, iterable, 1)
../pymbolic/pymbolic/mapper/evaluator.py:96: in <genexpr>
    return product(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
pytential/symbolic/execution.py:265: in map_common_subexpression
    rec = self.rec(expr.child)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
pytential/symbolic/execution.py:308: in map_call
    args = [self.rec(arg) for arg in expr.parameters]
pytential/symbolic/execution.py:308: in <listcomp>
    args = [self.rec(arg) for arg in expr.parameters]
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
../pymbolic/pymbolic/mapper/evaluator.py:92: in map_sum
    return sum(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/evaluator.py:92: in <genexpr>
    return sum(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
../pymbolic/pymbolic/mapper/evaluator.py:96: in map_product
    return product(self.rec(child) for child in expr.children)
../../miniconda3/lib/python3.6/site-packages/pytools/__init__.py:978: in product
    return reduce(mul, iterable, 1)
../pymbolic/pymbolic/mapper/evaluator.py:96: in <genexpr>
    return product(self.rec(child) for child in expr.children)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
pytential/symbolic/execution.py:265: in map_common_subexpression
    rec = self.rec(expr.child)
../pymbolic/pymbolic/mapper/__init__.py:134: in __call__
    return method(expr, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytential.symbolic.execution.EvaluationMapper object at 0x7f5dc4f33710>
expr = NumReferenceDerivative(((0, 1),), NodeCoordinateComponent(0, DOFDescriptor(geometry=point_target, stage=None, granularity=GRANULARITY_NODE)), DOFDescriptor(geometry=point_target, stage=None, granularity=GRANULARITY_NODE))

    def map_num_reference_derivative(self, expr):
        discr = self.places.get_discretization(
                expr.dofdesc.geometry, expr.dofdesc.discr_stage)
    
        from pytools import flatten
        ref_axes = flatten([axis] * mult for axis, mult in expr.ref_axes)
>       return discr.num_reference_derivative(ref_axes, self.rec(expr.operand))
E       AttributeError: 'PointsTarget' object has no attribute 'num_reference_derivative'

pytential/symbolic/execution.py:215: AttributeError
```